### ISSUE TITLE
Kick Off 

### ISSUE THREAD

**Issue by DClarkious** [SurvivalSim/LifeSim#1]
Let's talk about what we should do to modernise the survival simulator

**Comment by AidenGallagher** [comment]
I think we need a Minimum Viable Product first. So the minimum we need to do to have something that works. Which in my opinion is;
- map
- 1 predator
- 3 prey
- symbols to represent them
- random movement in any direction

Which should be pretty easy to do :)

**Comment by DClarkious** [comment]
I think we need to document what we have so far and then discuss requirements, the issues we faced before, we need to discuss technologies we want to use.

How far we want to get, should we go as far to implement some kind of natural language processing features that Dr Paul Bowden thought of in the beginning.

I really think we should really go forward with a test driven development strategy

**Comment by MasterGeneiJin** [comment]
I like Alex's suggestion of using Golang personally; it's a C like language with less complexity, comes built in code coverage tools just for you Dan. 

For the UI element [any of these libraries will do](https://github.com/avelino/awesome-go#gui) as long as it's cross platform, the qml one looks interesting personally.

**Comment by DClarkious** [comment]
I was thinking for the ui we go full 3d and ditch the 2d segmented world, that way we can simulate something akin to real life.

I think that going for golang is a good idea, definitely would be good to learn, add another string to our bows.

**Comment by DClarkious** [comment]
I think we should put the golang compiler (MSI and zip files) + write some batch scripts to set up the environmental variables necessary to run it, in the repo so we have the same starting point

**Comment by AlexRudd** [comment]
Hey, so what are we trying to achieve with this? Is it still trying to answer the question of whether communication improves survivability?

If it's a serious simulator we want to create then all we really care about is the outcome (simply how many iteration cycles the prey species survived with/without communication (possibly even broken down into frame states)).

Any visualisation would then be secondary. Either completely separate app that runs against the output of the simulator to create a visualisation you can then step through for fun; or something that polls/recieves the state of the simulation as it's running and creates visualisations in real-time.

If we want to investigate the NLP side of things do we create a real AI that uses some predefined lexicon to actually construct sentences that describe its situation and the things it wants to tell its friends, and also give it the ability to process and react to incoming arbitrary sentences?

We could mess around with machine learning or genetic algorithms to create AI's and then pit them against each other?

Basically is there something that everyone wants to get out of this, such as experience in a particular language, architecture, tools, or whatever?

**Comment by AlexRudd** [comment]
One of my workmates mentioned a uni project they did where they were given a game framework (tanks PvP) and then they each programmed a tank AI and then pitched their creations against each other.

We could do something similar where we use the predator prey sim as a way to have fun creating our own AI's and seeing how long we can get them to survive

**Comment by DClarkious** [comment]
Hey, I definitely agree that we should not even focus on the UI / visuals at all until the very end.

I think we should go the original route and test the hypothesis that "Communication improves survivability". I agree that we should create some kind of NLP engine that encodes and decodes messages that the prey will send.

I like the idea that we would incorporate a machine learning aspect to aid development of our AIs. 

For me I think the machine learning + NLP slant is what I want out of this project. even if it is dipping our toes into these topics

and I really like the idea that we turn this into a kind of game. perhaps we could have a high scores board to see who can create the most survivable / Vicious AI. Maybe even pitch it as some kind of Hunger Games scenario.

**Comment by AlexRudd** [comment]
Ooo I like the hunger games angle

**Comment by DClarkious** [comment]
Go seems like a good niche language to write this in.
The github repo that Michael shared does have a lot of the things that we have mentioned here to get us going.

**Comment by AlexRudd** [comment]
*to get us GOing
The best feature of the language is the potential for puns

It's a pretty huge language in the distributed computing and microservice world, we're slowly moving more and more of our platform to Go as we realise it's just better for what we do. If you've got some time, look up talks by Peter Bourgon or Dave Cheny and go through some of the tutorials. Channels, go routines, and implicit interfaces!

**Comment by AlexRudd** [comment]
Also wouldn't mind an excuse to try Rust. Saw a talk about it and seems neat
